java/arcs/core/data/expression/PaxelParser.kt

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data.expression
+
+import arcs.core.data.expression.Expression.FieldExpression
+import arcs.core.data.expression.Expression.QualifiedExpression
+import arcs.core.data.expression.Expression.Scope
+import arcs.core.util.AnyOfParser
+import arcs.core.util.ParseResult
+import arcs.core.util.Parser
+import arcs.core.util.ParserException
+import arcs.core.util.div
+import arcs.core.util.eof
+import arcs.core.util.many
+import arcs.core.util.map
+import arcs.core.util.optional
+import arcs.core.util.parser
+import arcs.core.util.plus
+import arcs.core.util.regex
+import arcs.core.util.token
+import arcs.core.util.unaryMinus
+import java.math.BigInteger
+
+/**
+ * A parser combinator implementation of the Paxel expression parser.
+ */
+object PaxelParser {
+    private val WS = "[\\s\\n]"
+
+    sealed class DiscreteType<T : Number> {
+        object PaxelBigInt : DiscreteType<BigInteger>() {
+            override fun parse(number: String) = BigInteger(number)
+        }
+
+        object PaxelInt : DiscreteType<Int>() {
+            override fun parse(number: String) = number.toInt()
+        }
+
+        object PaxelLong : DiscreteType<Long>() {
+            override fun parse(number: String) = number.toLong()
+        }
+
+        abstract fun parse(number: String): T
+    }
+
+    sealed class Unit(val conversionFactor: Int) {
+        fun convert(number: Number): Number = when (number) {
+            is BigInteger -> number.multiply(conversionFactor.toBigInteger())
+            is Long -> number * conversionFactor
+            is Double -> number * conversionFactor.toDouble()
+            else -> number.toLong() * conversionFactor
+        }
+
+        object Millisecond : Unit(1)
+        object Second : Unit(1000)
+        object Minute : Unit(Second.conversionFactor * 60)
+        object Hour : Unit(Minute.conversionFactor * 60)
+        object Day : Unit(Hour.conversionFactor * 24)
+    }
+
+    private val whitespace = -regex("($WS+)")
+
+    private val units =
+        optional(whitespace + (regex("(millisecond)s?").map { Unit.Millisecond } /
+        regex("(second)s?").map { Unit.Second } /
```

Comment:
brackets unnecessary here?

Reply:
If you remove the bracket, then the 'whitespace +' applies only to the first regex. We want a + (b or c or d), not a + b or c or d. Unfortunately, since I'm using '/' for 'or' and '+' for 'and', the precedence is reversed. 

Normally, yes, '+'  would ensure this.  I thought of using '*' for 'AND', and '+' for 'or', but '+' doesn't match the pegjs look-and-feel as much 👍

Reply:
To clarify, I was talking about the parens in the regexes (e.g. `"(millisecond)s?"` ) not the ones around the set of time unit options.

(Btw: Your note around precedence seems fine, though I think matching pegjs is not actually desirable, I don't think anyone is currently arguing that it should be our long term solution for typescript parsing, just that we will keep it until the switching cost is higher. Thanks though, I do appreciate your attention to detail :) )